more_itertools/recipes.py

```diff
@@ -1383,6 +1365,72 @@ def _running_median_minheap_only(iterable):  # pragma: no cover
             yield (hi[0] - lo[0]) / 2
 
 
-if not _max_heap_available:
-    _running_median_minheap_only.__doc__ = running_median.__doc__
-    running_median = _running_median_minheap_only
+def _sorted_window(iterator, k):  # pragma: no cover
+    "Yield windows in sorted order"
+
+    # This simple implementation is reasonably fast with two O(1) deque
+    # updates and two O(log k) data comparisons per iteration.  However,
+    # it also has two O(k) steps, a list insertion and a list deletion.
+
+    # Grant Jenks' work on SortedCollections showed that those two steps
+    # have a very low constant factor because they are implemented with
+    # highly optimized C memmoves that shift the data pointers without
+    # needing to touch the underlying data with reference counts.
+
+    # This function could be reimplemented with SortedCollections, blist
+    # some other binary tree, or an IndexableSkipList all of which have
+    # O(k) insertions and deletions, albiet with a larger constant
+    # factor. For very large window sizes, this might matter.
+
+    history = deque()
+    window = []
+    for x in iterator:
+        history.append(x)
+        insort_left(window, x)
+        if len(window) > k:
+            i = bisect_left(window, history.popleft())
+            del window[i]
+        yield window
+
+
+def _running_median_windowed(iterable, window):
+    "Yield median of values in a sliding window."
+
+    for data in _sorted_window(iterable, k=window):
+        n = len(data)
+        if n % 2 == 1:
+            yield data[n // 2]
+        else:
+            i = n // 2
+            yield (data[i - 1] + data[i]) / 2
+
+
+def running_median(iterable, *, window=None):  # pragma: no cover
+    """Cumulative median of values seen so far or values in a sliding window.
+
+    For example:
+
+        >>> list(running_median([5.0, 9.0, 4.0, 12.0, 8.0, 9.0]))
+        [5.0, 7.0, 5.0, 7.0, 8.0, 8.5]
+        >>> list(running_median([5.0, 9.0, 4.0, 12.0, 8.0, 9.0], window=3))
+        [5.0, 7.0, 5.0, 9.0, 8.0, 9.0]
+
+    Supports numeric types such as int, float, Decimal, and Fraction,
+    but not complex numbers which are unorderable.
+
+    On version Python 3.13 and prior, max-heaps are simulated with
+    negative values. The negation causes Decimal inputs to apply context
+    rounding, making the results slightly different than that obtained
+    by statistics.median().
+    """
+
+    iterator = iter(iterable)
+
+    if window is not None:
+        window = index(window)
```

Comment:
Why `index()` here?

Reply:
Perhaps there should also be a test to reject non-positive window sizes.

Reply:
Should we be using it on `islice_extended`?

Reply:
The checks are already in the downstream calls, so incorrect arguments are getting rejected. However, that only occurs after the generator starts running. Ideally, the arguments would be checked on instantiation, but I think `islice_extended` is probably fine as-is and it hasn't caused any known problems in the real world.

```python
>>> iterator = islice_extended('abcdefgh', -4.0, -1)  # Error not detected on instantiation
>>> list(iterator)  # Error detected when run
...
TypeError: an integer is required
```